src/Pester.Runtime.ps1

```diff
@@ -2170,6 +2187,24 @@ function PostProcess-DiscoveredBlock {
         }
 
         $b.ShouldRun = $shouldRunBasedOnChildren
+
+        if ($b.ShouldRun) {
+            if (-not $b.Skip -and $shouldSkipBasedOnChildren) {
+                if ($PesterPreference.Debug.WriteDebugMessages.Value) {
+                    if ($b.IsRoot) {
+                        Write-PesterDebugMessage -Scope Skip "($($b.BlockContainer)) Container will be skipped because all children are excluded or skipped."
+                    } else {
+                        Write-PesterDebugMessage -Scope Skip "($($b.Path -join '.')) Block will be skipped because all children are excluded or skipped."
+                    }
+                }
+                $b.Skip = $true
+            } elseif ($b.Skip -and -not $shouldSkipBasedOnChildren) {
+                if ($PesterPreference.Debug.WriteDebugMessages.Value) {
+                    Write-PesterDebugMessage -Scope Skip "($($b.Path -join '.')) Block was marked as Skip, but one or more children should run and is not skipped, so the block will not be skipped."
+                }
+                $b.Skip = $false
+            }
```

Comment:
Why did I add this again? 😅

Reply:
Oh, it was to make BeforeAll/AfterAll run for blocks with `-Skip` if a child test will run due to Explicit include (line filter).

Pester 5.5.0

```powershell
Filter 'Line' set to ('C:\Users\Frode\Git\Pester/demoBlockSkip.tests.ps1:37').
Filters selected 1 tests to run.
Discovery: Test discovery finished.
Running tests.

Running tests from 'C:\Users\Frode\Git\Pester\demoBlockSkip.tests.ps1'
Before ALL ROOT
Describing A
 Describing B
Before EACH -A
Before EACH -B
This is test 2
After EACH -B
After EACH -A
   [+] Test 2 6ms (4ms|2ms)
After ALL ROOT
Tests completed in 72ms
Tests Passed: 1, Failed: 0, Skipped: 0, Inconclusive: 0, NotRun: 3
```

PR as-is

```powershell
Filter 'Line' set to ('/workspaces/Pester/demoBlockSkip.tests.ps1:37').
Filters selected 1 tests to run.
Discovery: Test discovery finished. 
Running tests.

Running tests from '/workspaces/Pester/demoBlockSkip.tests.ps1'
Before ALL ROOT
Before ALL -A
Before ALL -B
Describing A
 Describing B
Before EACH -A
Before EACH -B
This is test 2
After EACH -B
After EACH -A
   [+] Test 2 31ms (28ms|2ms)
After ALL -B
After ALL -A
After ALL ROOT
Tests completed in 145ms
Tests Passed: 1, Failed: 0, Skipped: 0 NotRun: 3
```

PR without the elseif-block (don't undo Skip):

```powershell
Filter 'Line' set to ('/workspaces/Pester/demoBlockSkip.tests.ps1:37').
Filters selected 1 tests to run.
Discovery: Test discovery finished. 
Running tests.

Running tests from '/workspaces/Pester/demoBlockSkip.tests.ps1'
Describing A
 Describing B
Before EACH -A
Before EACH -B
This is test 2
After EACH -B
After EACH -A
   [+] Test 2 53ms (25ms|28ms)
Tests completed in 413ms
Tests Passed: 1, Failed: 0, Skipped: 0 NotRun: 3
```

Which scenario is correct?

Reply:
Yes, that was the idea. Test is broken you skip it, but you can still run it if you run just that. Another alternative is using a tag, but then you have to always exclude that tag.

Reply:
Cool. I'll keep the fix to run all before-/afterall blocks 👍

We need to get this logic documented before v6.